File: 0036-valid-sudoku/0036-valid-sudoku.ts
function isValidSudoku(board: string[][]): boolean {
  const set = new Set<number>();
  for (let row = 0; row < board.length; row++) {
    for (let col = 0; col < board.length; col++) {
      const val = Number(board[row][col]);
      if (!Number.isNaN(val)) {
        const box = (row / 3 | 0) * 3 + (col / 3 | 0);
        if (
          set.has(row + 1 << 4 | val) ||
          set.has(col + 1 << 8 | val) ||
          set.has(box + 1 << 12 | val)
        ) return false;
        set.add(row + 1 << 4 | val);
        set.add(col + 1 << 8 | val);
        set.add(box + 1 << 12 | val);
      }
    } 
  } 
  return true;
};

File: 0036-valid-sudoku/0036-valid-sudoku.go
import (
// "strconv"
)

func isValidSudoku(board [][]byte) bool {
	set := map[int]bool{}
	for row := 0; row < len(board); row++ {
		for col := 0; col < len(board); col++ {
			val := int(rune(board[row][col]) - '0')
			if val != -2 {
				box := ((row / 3) * 3) + col/3
				if set[(row+1)<<4|val] || set[(col+1)<<8|val] || set[(box+1)<<12|val] {
					return false
				}
				set[(row+1)<<4|val] = true
				set[(col+1)<<8|val] = true
				set[(box+1)<<12|val] = true
			}
		}
	}
	return true
}